### PR TITLE
HDDS-4908. Layout version should be available in DB for an un-finalized OM to be finalized through a Ratis snapshot.

### PR DESCRIPTION
**Pull request by avijayanhwx**
## What changes were proposed in this pull request?

Although OM Finalization is a Ratis request, it updates a persistent location outside the OM state machine maintained "RocksDB". Hence, in a rare case where an unfinalized OM needs a Ratis snapshot to finalize (without any logs), the VERSION file will not be updated, and hence the OM will be stuck in that state.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-4908

## How was this patch tested?
Manually tested by adding a new layout feature.
Added integration tests.

### CODE REVIEW

**Review by errose28 — COMMENTED** [review]
Thanks for working on this @avijayanhwx. A few comments left in line.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/upgrade/OMUpgradeFinalizer.java [hunk 1]
```diff
@@ -100,6 +104,7 @@ public Void call() throws IOException {
         for (OMLayoutFeature f : versionManager.unfinalizedFeatures()) {
           finalizeFeature(f);
           updateLayoutVersionInVersionFile(f, ozoneManager.getOmStorage());
+          updateLayoutVersionInDB(f, ozoneManager);
```
[the remark this answers is not in the document]
Comment (avijayanhwx) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/upgrade/OMUpgradeFinalizer.java]: Agree, I have changed the way we write to the DB.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 2]
```diff
@@ -550,6 +551,23 @@ private void instantiateServices(boolean withNewSnapshot) throws IOException {
...
+    if (withNewSnapshot) {
+      Integer layoutVersionInDB = getLayoutVersionInDB();
+      if (layoutVersionInDB != null &&
+          versionManager.getMetadataLayoutVersion() < layoutVersionInDB) {
+        LOG.info("New OM snapshot received with higher layout version {}. " +
+            "Attempting to finalize current OM to that version.",
+            layoutVersionInDB);
+        upgradeFinalizer.finalizeAndWaitForCompletion(
+            "om-ratis-snapshot", this,
+            TimeUnit.SECONDS.toSeconds(10));
```
[the remark this answers is not in the document]
Comment (avijayanhwx) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java]: Made it configurable.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java [hunk 3]
```diff
@@ -143,6 +143,7 @@
   public static final String PREFIX_TABLE = "prefixTable";
   public static final String TRANSACTION_INFO_TABLE =
       "transactionInfoTable";
+  public static final String META_TABLE = "metaTable";
```
[the remark this answers is not in the document]
Comment (avijayanhwx) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java:146]: Using this table as a generic table to store String,String "meta" information.

**Comment by errose28** [comment]
Thanks @avijayanhwx LGTM +1 pending CI.

**Review by swagle — APPROVED** [review]

**Comment by avijayanhwx** [comment]
Thanks for the review @errose28 & @swagle.

**Pull request merged**